test/e2e/storage/csi_mock_volume.go

```diff
@@ -1165,6 +1208,94 @@ var _ = utils.SIGDescribe("CSI mock volume", func() {
 			})
 		}
 	})
+
+	// These tests *only* work on a cluster which has the CSIVolumeFSGroupPolicy feature enabled.
+	ginkgo.Context("CSI FSGroupPolicy", func() {
+		tests := []struct {
+			name          string
+			fsGroupPolicy storagev1.FSGroupPolicy
+		}{
+			{
+				name:          "should modify fsGroup if fsGroupPolicy=File",
+				fsGroupPolicy: storagev1.FileFSGroupPolicy,
+			},
+			{
+				name:          "should not modify fsGroup if fsGroupPolicy=None",
+				fsGroupPolicy: storagev1.NoneFSGroupPolicy,
+			},
+		}
+		for _, t := range tests {
+			test := t
+			ginkgo.It(test.name, func() {
+				if framework.NodeOSDistroIs("windows") {
+					e2eskipper.Skipf("FSGroupPolicy is only applied on linux nodes -- skipping")
+				}
+				init(testParameters{
+					disableAttach:  true,
+					registerDriver: true,
+					fsGroupPolicy:  test.fsGroupPolicy,
+				})
+				defer cleanup()
+
+				syncDelay := 5 * time.Second
+
+				fsGroupVal := int64(rand.Int63n(20000) + 1024)
+				fsGroup := &fsGroupVal
+
+				_, _, pod := createPodWithFSGroup(fsGroup) /* persistent volume */
+
+				mountPath := pod.Spec.Containers[0].VolumeMounts[0].MountPath
+				fileName := mountPath + "/" + f.UniqueName
+
+				waitCtx, cancel := context.WithTimeout(context.Background(), podStartTimeout)
+				defer cancel()
+				condition := anyOf(
+					podRunning(waitCtx, f.ClientSet, pod.Name, pod.Namespace),
+					// We only just created the CSIDriver objects, therefore
+					// we have to ignore all older events, plus the syncDelay as our
+					// safety margin.
+					podHasStorage(waitCtx, f.ClientSet, pod.Name, pod.Namespace, time.Now().Add(syncDelay)),
+				)
+				err := wait.PollImmediateUntil(poll, condition, waitCtx.Done())
+				framework.ExpectNoError(err, "failed to start pod")
+
+				// Inject the contents onto the mount
+				tk := e2ekubectl.NewTestKubeconfig(framework.TestContext.CertDir, framework.TestContext.Host, framework.TestContext.KubeConfig, framework.TestContext.KubeContext, framework.TestContext.KubectlPath, pod.Namespace)
+				err = tk.WriteFileViaContainer(pod.Name, pod.Spec.Containers[0].Name, fileName, "filecontents")
+				framework.ExpectNoError(err, "failed: writing the contents: %s", err)
+
+				// Delete the created file. This step is mandatory, as the mock driver
+				// won't clean up the contents automatically.
+				defer func() {
+					delete := []string{"exec", pod.Name, fmt.Sprintf("--namespace=%v", pod.Namespace), "--"}
+					delete = append(delete, "rm", "-f", fileName)
+					_, err = framework.RunKubectl(pod.Namespace, delete...)
+					framework.ExpectNoError(err, "failed: deleting the file: %s", err)
+				}()
+
+				// Get the fsGroup from the created file
```

Comment:
Is it possible we could have a race where the Pod gets created before kubelet sees the CSIDriver object?